src/back/spv/writer.rs

```diff
@@ -1214,6 +1214,41 @@ impl Writer {
                         },
                         _ => unreachable!(),
                     },
+                    crate::BinaryOperator::And => match *left_ty_inner {
+                        crate::TypeInner::Vector { .. } => match *right_ty_inner {
```

Comment:
Here and in other binary operators we don't actually need to match the left and right types, at all. The valid IR guarantees that they are the same, and they are applicable. You just need to generate the proper SPIR-V instruction with that in mind, no need to re-validate the types.
The only places where types need to be checked are places where instructions are different, e.g. multiply instructions f_add/i_add, etc

Reply:
Yes that is true! This is the first binary operator that uses one instruction for all its result types.